src/optimizer/cardinality_estimator.cpp

```diff
@@ -159,28 +152,46 @@ void CardinalityEstimator::EstimateCardinality(JoinNode *node) {
 	idx_t tdom_join_left = GetTDom(node->left_binding);
 	D_ASSERT(tdom_join_left == tdom_join_right);
 #endif
-	auto left_card = node->left->cardinality;
-	auto right_card = node->right->cardinality;
+	auto left_card = node->left->estimated_props->cardinality;
+	auto right_card = node->right->estimated_props->cardinality;
 	D_ASSERT(tdom_join_right != 0);
 	D_ASSERT(tdom_join_right != NumericLimits<idx_t>::Maximum());
 	node->cardinality = (left_card * right_card) / tdom_join_right;
+	node->estimated_props->cardinality = node->cardinality;
+}
+
+static bool IsLogicalFilter(LogicalOperator *op) {
+	return op->type == LogicalOperatorType::LOGICAL_FILTER;
 }
 
+static bool IsLogicalGet(LogicalOperator *op) {
+	return op->type == LogicalOperatorType::LOGICAL_GET;
+}
 
-static LogicalGet* GetLogicalGet(LogicalOperator *op) {
+static LogicalGet *GetLogicalGet(LogicalOperator *op) {
 	LogicalGet *get = nullptr;
-	if (op->type == LogicalOperatorType::LOGICAL_GET) {
-		get = (LogicalGet*)op;
-	}
-	if (op->type == LogicalOperatorType::LOGICAL_FILTER) {
-		if (op->children[0]->type == LogicalOperatorType::LOGICAL_GET) {
-			get = (LogicalGet*)op->children.at(0).get();
+	if (IsLogicalGet(op)) {
```

Comment:
Seems like this code would be cleaner with a switch statement:
```c++
switch (op->type) {
    case LogicalOperatorType::LOGICAL_GET: {
        // ...
    }
    case LogicalOperatorType::LOGICAL_FILTER: {
        // ...
    }
}
```

Then you don't need the `IsLogicalGet` function, and it's easy to extend with more operator types.